File: luajit/tests/TestDeviceAPI.lua
-- Copyright (c) 2021-2022 Nicholas Corgan
-- SPDX-License-Identifier: BSL-1.0

SoapySDR = require("SoapySDR")

bit = require("bit")
ffi = require("ffi")
luaunit = require("luaunit")

function testArgInfo()
    -- We can only get a meaningful ArgInfo from something that actually
    -- returns one, so take advantage of the loopback module if we have it.
    local dev = nil

    -- For the sake of pcall
    local function makeDev()
        dev = SoapySDR.Device.new("driver=loopback")
    end

    -- Hopefully this doesn't change.
    -- https://github.com/JuliaTelecom/SoapyLoopback/blob/main/Streaming.cpp#L51
    if pcall(makeDev) then
        local argInfoList = dev:getStreamArgsInfo(SoapySDR.Direction.RX, 0)
        luaunit.assertEquals(#argInfoList, 3)

        for i = 1,3 do
            luaunit.assertIsTable(argInfoList[i])
            luaunit.assertIsString(argInfoList[i].key)
            luaunit.assertIsNumber(argInfoList[i].value)
            luaunit.assertIsString(argInfoList[i].name)
            luaunit.assertIsString(argInfoList[i].description)
            luaunit.assertIsString(argInfoList[i].units)
            luaunit.assertEquals(argInfoList[i].argType, SoapySDR.ArgInfoType.INT)
            luaunit.assertIsTable(argInfoList[i].range)
            luaunit.assertIsNumber(argInfoList[i].range.minimum)
            luaunit.assertIsNumber(argInfoList[i].range.maximum)
            luaunit.assertIsNumber(argInfoList[i].range.step)
            luaunit.assertIsTable(argInfoList[i].options)
            luaunit.assertIsTable(argInfoList[i].optionNames)
        end
    else
        print "Loopback module not found. Skipping test."
    end
end

local function testDeviceWithDirection(device, direction)
    -- For the null device, functions generally don't do anything or return some
    -- hardcoded value, but we can still make sure the LuaJIT translates values
    -- back and forth as expected.

    --
    -- Channels API
    --
    device:setFrontendMapping(direction, "0:0")
    device:setFrontendMapping(direction, "0:0")
    luaunit.assertIsString(device:getFrontendMapping(direction))
    luaunit.assertIsTable(device:getChannelInfo(direction, 0))
    luaunit.assertIsBoolean(device:getFullDuplex(direction, 0))

    --
    -- Stream API
    --
    luaunit.assertIsTable(device:getStreamFormats(direction, 0))
    local format, fullScale = unpack(device:getNativeStreamFormat(direction, 0))
    luaunit.assertEquals(format, SoapySDR.Format.CS16)
    luaunit.assertEquals(fullScale, bit.lshift(1, 15))
    luaunit.assertIsTable(device:getStreamArgsInfo(direction, 0))

    local format = SoapySDR.Format.CF32
    local channels = {0,1}
    local args = "bufflen=8192,buffers=15"
    local stream = device:setupStream(direction, format, channels) -- Without optional param
    stream = device:setupStream(direction, format, channels, args) -- With optional param

    luaunit.assertEquals(device:getStreamMTU(stream), 1024)

    local flags = bit.bor(SoapySDR.StreamFlags.HAS_TIME, SoapySDR.StreamFlags.END_BURST)
    local timeNs = 1000
    local numElems = 1024
    luaunit.assertEquals(
        device:activateStream(stream, flags, timeNs, numElems),
        SoapySDR.Error.NOT_SUPPORTED)
    -- Default parameter causes specific null device behavior
    luaunit.assertEquals(device:activateStream(stream), 0)

    local cf32Buff = ffi.new("complex float[?]", numElems)
    local cf32Buff2D = ffi.new("complex float*[1]", {cf32Buff})
    local timeoutUs = 100000

    local readOutput = device:readStream(stream, cf32Buff2D, numElems, timeoutUs)
    luaunit.assertEquals(readOutput[1], SoapySDR.Error.NOT_SUPPORTED)
    luaunit.assertEquals(readOutput[2], 0)
    luaunit.assertEquals(readOutput[3], 0)

    readOutput = device:readStream(stream, cf32Buff2D, numElems) -- Without optional parameter
    luaunit.assertEquals(readOutput[1], SoapySDR.Error.NOT_SUPPORTED)
    luaunit.assertEquals(readOutput[2], 0)
    luaunit.assertEquals(readOutput[3], 0)

    local writeOutput = device:writeStream(stream, cf32Buff2D, numElems, flags, timeNs, timeoutUs)
    luaunit.assertEquals(writeOutput[1], SoapySDR.Error.NOT_SUPPORTED)
    luaunit.assertEquals(writeOutput[2], flags)

    writeOutput = device:writeStream(stream, cf32Buff2D, numElems) -- Without optional parameters
    luaunit.assertEquals(writeOutput[1], SoapySDR.Error.NOT_SUPPORTED)
    luaunit.assertEquals(writeOutput[2], 0)

    local readStreamStatusOutput = device:readStreamStatus(stream, timeoutUs)
    luaunit.assertEquals(readStreamStatusOutput[1], SoapySDR.Error.NOT_SUPPORTED)
    luaunit.assertEquals(readStreamStatusOutput[2], 0)
    luaunit.assertEquals(readStreamStatusOutput[3], 0)
    luaunit.assertEquals(readStreamStatusOutput[4], 0)

    local readStreamStatusOutput = device:readStreamStatus(stream) -- Without optional parameter
    luaunit.assertEquals(readStreamStatusOutput[1], SoapySDR.Error.NOT_SUPPORTED)
    luaunit.assertEquals(readStreamStatusOutput[2], 0)
    luaunit.assertEquals(readStreamStatusOutput[3], 0)
    luaunit.assertEquals(readStreamStatusOutput[4], 0)

    luaunit.assertEquals(
        device:deactivateStream(stream, flags, timeNs),
        SoapySDR.Error.NOT_SUPPORTED)
    luaunit.assertEquals(device:closeStream(stream), 0)

    --
    -- Antenna API
    --
    luaunit.assertIsTable(device:listAntennas(direction, 0))
    device:setAntenna(direction, 0, "ANT")
    luaunit.assertIsString(device:getAntenna(direction, 0))

    --
    -- Frontend corrections API
    --
    luaunit.assertIsBoolean(device:hasDCOffsetMode(direction, 0))
    device:setDCOffsetMode(direction, 0, true)
    luaunit.assertIsBoolean(device:getDCOffsetMode(direction, 0))
    luaunit.assertIsBoolean(device:hasDCOffset(direction, 0))

    -- Test all valid types
    device:setDCOffset(direction, 0, ffi.new("complex", 1.0, 0.0))
    device:setDCOffset(direction, 0, ffi.new("complex float", 1.0, 0.0))
    device:setDCOffset(direction, 0, 1.0)
    luaunit.assertEquals(
        ffi.typeof(device:getDCOffset(direction, 0)),
        ffi.typeof("complex"))

    luaunit.assertIsBoolean(device:hasIQBalance(direction, 0))

    -- Test all valid types
    device:setIQBalance(direction, 0, ffi.new("complex", 1.0, 0.0))
    device:setIQBalance(direction, 0, ffi.new("complex float", 1.0, 0.0))
    device:setIQBalance(direction, 0, 1.0)
    luaunit.assertEquals(
        ffi.typeof(device:getIQBalance(direction, 0)),
        ffi.typeof("complex"))

    luaunit.assertIsBoolean(device:hasIQBalanceMode(direction, 0))
    device:setIQBalanceMode(direction, 0, true)
    luaunit.assertIsBoolean(device:getIQBalanceMode(direction, 0))

    luaunit.assertIsBoolean(device:hasFrequencyCorrection(direction, 0))
    device:setFrequencyCorrection(direction, 0, 0.0)
    luaunit.assertIsNumber(device:getFrequencyCorrection(direction, 0))

    --
    -- Gain API
    --
    luaunit.assertIsTable(device:listGains(direction, 0))
    luaunit.assertIsBoolean(device:hasGainMode(direction, 0))
    device:setGainMode(direction, 0, true)
    luaunit.assertIsBoolean(device:getGainMode(direction, 0))

    device:setGain(direction, 0, 0.0)
    device:setGainElement(direction, 0, "", 0.0)
    luaunit.assertIsNumber(device:getGain(direction, 0))
    luaunit.assertIsNumber(device:getGainElement(direction, 0, ""))

    luaunit.assertIsTable(device:getGainRange(direction, 0))
    luaunit.assertIsTable(device:getGainElementRange(direction, 0, ""))

    --
    -- Frequency API
    --
    device:setFrequency(direction, 0, 0.0)
    device:setFrequency(direction, 0, 0.0, {})
    device:setFrequency(direction, 0, 0.0, "")

    device:setFrequencyComponent(direction, 0, "", 0.0)
    device:setFrequencyComponent(direction, 0, "", 0.0, {})
    device:setFrequencyComponent(direction, 0, "", 0.0, "")

    luaunit.assertEquals(device:getFrequency(direction, 0), 0.0)
    luaunit.assertEquals(device:getFrequencyComponent(direction, 0, ""), 0.0)

    luaunit.assertIsTable(device:listFrequencies(direction, 0))

    luaunit.assertIsTable(device:getFrequencyRange(direction, 0))
    luaunit.assertIsTable(device:getFrequencyRangeComponent(direction, 0, ""))
    luaunit.assertIsTable(device:getFrequencyArgsInfo(direction, 0))

    --
    -- Sample rate API
    --
    device:setSampleRate(direction, 0, 0.0)
    luaunit.assertEquals(device:getSampleRate(direction, 0), 0.0)
    luaunit.assertIsTable(device:getSampleRateRange(direction, 0))

    --
    -- Bandwidth API
    --
    device:setBandwidth(direction, 0, 0.0)
    luaunit.assertEquals(device:getBandwidth(direction, 0), 0.0)
    luaunit.assertIsTable(device:getBandwidthRange(direction, 0))

    --
    -- Sensor API
    --
    luaunit.assertIsTable(device:listChannelSensors(direction, 0))
    luaunit.assertIsTable(device:getChannelSensorInfo(direction, 0, ""))
    luaunit.assertIsString(device:readChannelSensor(direction, 0, ""))

    --
    -- Settings API
    --
    luaunit.assertIsTable(device:getChannelSettingInfo(direction, 0))
    luaunit.assertIsTable(device:getChannelSettingInfoWithKey(direction, 0, ""))
    device:writeChannelSetting(direction, 0, "", "")
    luaunit.assertIsString(device:readChannelSetting(direction, 0, ""))
end

function testDevice()
    -- Make sure either method works.
    local device = SoapySDR.Device.new("driver=null,type=null")
    luaunit.assertEquals(tostring(device), "null:null")
    device = SoapySDR.Device.new({driver="null",type="null"})
    luaunit.assertEquals(tostring(device), "null:null")

    --
    -- Identification API
    --
    luaunit.assertEquals("null", device:getDriverKey())
    luaunit.assertEquals("null", device:getHardwareKey())
    luaunit.assertIsTable(device:getHardwareInfo())

    --
    -- Clocking API
    --
    device:setMasterClockRate(0.0)
    luaunit.assertEquals(device:getMasterClockRate(), 0.0)
    luaunit.assertIsTable(device:getMasterClockRates())

    device:setReferenceClockRate(0.0)
    luaunit.assertEquals(device:getReferenceClockRate(), 0.0)
    luaunit.assertIsTable(device:getReferenceClockRates())

    luaunit.assertIsTable(device:listClockSources())
    device:setClockSource("")
    luaunit.assertEquals(device:getClockSource(), "")

    --
    -- Time API
    --
    luaunit.assertIsTable(device:listTimeSources())
    device:setTimeSource("")
    luaunit.assertEquals(device:getTimeSource(), "")

    luaunit.assertIsBoolean(device:hasHardwareTime())
    luaunit.assertIsBoolean(device:hasHardwareTime(""))
    luaunit.assertEquals(device:getHardwareTime(), 0)
    luaunit.assertEquals(device:getHardwareTime(""), 0)
    device:setHardwareTime(0)
    device:setHardwareTime(0, "")

    --
    -- Sensor API
    --
    luaunit.assertIsTable(device:listSensors())
    luaunit.assertIsTable(device:getSensorInfo(""))
    luaunit.assertEquals(device:readSensor(""), "")

    --
    -- Register API
    --
    luaunit.assertIsTable(device:listRegisterInterfaces())

    device:writeRegister("", 0, 0)
    luaunit.assertEquals(device:readRegister("", 0), 0)

    device:writeRegisters("", 0, {})
    luaunit.assertIsTable(device:readRegisters("", 0, 0))

    --
    -- Settings API
    --
    luaunit.assertIsTable(device:getSettingInfo())
    luaunit.assertIsTable(device:getSettingInfoWithKey(""))
    device:writeSetting("", "")
    luaunit.assertEquals(device:readSetting(""), "")

    --
    -- GPIO API
    --
    luaunit.assertIsTable(device:listGPIOBanks())
    device:writeGPIO("", 0)
    device:writeGPIOMasked("", 0, 0)
    luaunit.assertEquals(device:readGPIO(""), 0)
    device:writeGPIODir("", 0)
    device:writeGPIODirMasked("", 0, 0)
    luaunit.assertEquals(device:readGPIODir(""), 0)

    --
    -- I2C API
    --
    device:writeI2C(0, "")
    luaunit.assertEquals(device:readI2C(0, 0), "")

    --
    -- SPI API
    --
    luaunit.assertEquals(device:transactSPI(0, 0, 0), 0)

    --
    -- UART API
    --
    luaunit.assertIsTable(device:listUARTs())
    device:writeUART("", "")
    luaunit.assertEquals(device:readUART(""), "")

    --
    -- Test per-direction API with both directions
    --
    testDeviceWithDirection(device, SoapySDR.Direction.TX)
    testDeviceWithDirection(device, SoapySDR.Direction.RX)
end

local runner = luaunit.LuaUnit.new()
os.exit(runner:runSuite())

File: swig/csharp/tests/TestDeviceAPI.cs
// Copyright (c) 2022 Nicholas Corgan
// SPDX-License-Identifier: BSL-1.0

using System.Collections.Generic;

using NUnit.Framework;

// For the null device, functions generally don't do anything or return some
// hardcoded value, but we can still make sure functions compile and run as
// expected, especially for the added C# fanciness.

[TestFixture]
public class TestDeviceAPI
{
    private void TestDeviceKeys(Pothosware.SoapySDR.Device device)
    {
        Assert.AreEqual("null", device.DriverKey);
        Assert.AreEqual("null", device.HardwareKey);
        Assert.AreEqual("null:null", device.ToString());
    }

    private Pothosware.SoapySDR.Device GetTestDevice()
    {
        // Make sure either method works.
        TestDeviceKeys(new Pothosware.SoapySDR.Device("driver=null,type=null"));

        var args = new Dictionary<string, string>();
        args.Add("driver", "null");
        args.Add("type", "null");

        var device = new Pothosware.SoapySDR.Device(args);
        TestDeviceKeys(device);

        return device;
    }

    [Test]
    public void Test_DeviceGeneralFunctions()
    {
        var device = GetTestDevice();

        //
        // Clocking API
        //

        device.MasterClockRate = 0.0;
        Assert.AreEqual(0.0, device.MasterClockRate);
        _ = device.MasterClockRates;

        device.ReferenceClockRate = 0.0;
        Assert.AreEqual(0.0, device.ReferenceClockRate);
        _ = device.ReferenceClockRates;

        device.ClockSource = "";
        Assert.AreEqual("", device.ClockSource);
        _ = device.ClockSources;

        //
        // Time API
        //

        device.TimeSource = "";
        Assert.AreEqual("", device.TimeSource);
        _ = device.TimeSources;

        _ = device.HasHardwareTime("");
        Assert.AreEqual(0, device.GetHardwareTime(""));
        device.SetHardwareTime(0);
        device.SetHardwareTime(0, "");

        //
        // Sensor API
        //

        _ = device.ListSensors();
        _ = device.GetSensorInfo("");

        _ = device.ReadSensor("");
        _ = device.ReadSensor<bool>("");
        /*
        _ = device.ReadSensor<short>("");
        _ = device.ReadSensor<int>("");
        _ = device.ReadSensor<long>("");
        _ = device.ReadSensor<ushort>("");
        _ = device.ReadSensor<uint>("");
        _ = device.ReadSensor<ulong>("");
        _ = device.ReadSensor<float>("");
        _ = device.ReadSensor<double>("");
        */
        _ = device.ReadSensor<string>("");

        //
        // Register API
        //

        _ = device.RegisterInterfaces;

        device.WriteRegister("", 0, 0);
        Assert.AreEqual(0, device.ReadRegister("", 0));

        device.WriteRegisters("", 0, new uint[] { 0, 0, 0 });
        _ = device.ReadRegisters("", 0, 1);

        //
        // Settings API
        //

        _ = device.GetSettingInfo();
        _ = device.GetSettingInfo("");

        device.WriteSetting("", 0);
        device.WriteSetting("", 0.0);
        device.WriteSetting("", false);
        device.WriteSetting("", "");

        /*
        _ = device.ReadSetting("");
        _ = device.ReadSetting<bool>("");
        
        _ = device.ReadSetting<short>("");
        _ = device.ReadSetting<int>("");
        _ = device.ReadSetting<long>("");
        _ = device.ReadSetting<ushort>("");
        _ = device.ReadSetting<uint>("");
        _ = device.ReadSetting<ulong>("");
        _ = device.ReadSetting<float>("");
        _ = device.ReadSetting<double>("");
        
        _ = device.ReadSetting<string>("");
        */

        //
        // GPIO API
        //

        _ = device.GPIOBanks;

        device.WriteGPIO("", 0);
        device.WriteGPIO("", 0, 0);
        Assert.AreEqual(0, device.ReadGPIO(""));

        device.WriteGPIODir("", 0);
        device.WriteGPIODir("", 0, 0);
        Assert.AreEqual(0, device.ReadGPIODir(""));

        //
        // I2C API
        //

        device.WriteI2C(0, "");
        Assert.AreEqual("", device.ReadI2C(0, 0));

        //
        // SPI API
        //

        Assert.AreEqual(0, device.TransactSPI(0, 0, 0));

        //
        // UART API
        //

        _ = device.UARTs;
        device.WriteUART("", "");
        Assert.AreEqual("", device.ReadUART("", 1000));
    }

    [Test]
    [TestCase(Pothosware.SoapySDR.Direction.Rx)]
    [TestCase(Pothosware.SoapySDR.Direction.Tx)]
    public void Test_DeviceNonStreamingDirectionFunctions(Pothosware.SoapySDR.Direction direction)
    {
        var device = GetTestDevice();

        //
        // Channels API
        //

        device.SetFrontendMapping(direction, "");
        _ = device.GetFrontendMapping(direction);
        _ = device.GetNumChannels(direction);
        _ = device.GetChannelInfo(direction, 0);
        _ = device.GetFullDuplex(direction, 0);

        //
        // Stream API
        //

        _ = device.GetStreamFormats(direction, 0);

        double fullScale;
        Assert.AreEqual(Pothosware.SoapySDR.StreamFormat.ComplexInt16, device.GetNativeStreamFormat(direction, 0, out fullScale));
        Assert.AreEqual(1 << 15, fullScale);

        _ = device.GetStreamArgsInfo(direction, 0);

        //
        // Antenna API
        //

        _ = device.ListAntennas(direction, 0);
        _ = device.GetAntenna(direction, 0);
        device.SetAntenna(direction, 0, "ANT");

        //
        // Frontend corrections API
        //

        _ = device.HasDCOffsetMode(direction, 0);
        _ = device.GetDCOffsetMode(direction, 0);
        device.SetDCOffsetMode(direction, 0, true);

        _ = device.HasDCOffset(direction, 0);
        _ = device.GetDCOffset(direction, 0);
        device.SetDCOffset(direction, 0, new System.Numerics.Complex(1.0, 0.0));

        _ = device.HasIQBalanceMode(direction, 0);
        _ = device.GetIQBalanceMode(direction, 0);
        device.SetIQBalanceMode(direction, 0, true);

        _ = device.HasIQBalance(direction, 0);
        _ = device.GetIQBalance(direction, 0);
        device.SetIQBalance(direction, 0, new System.Numerics.Complex(1.0, 0.0));

        _ = device.HasFrequencyCorrection(direction, 0);
        _ = device.GetFrequencyCorrection(direction, 0);
        device.SetFrequencyCorrection(direction, 0, 0.0);

        //
        // Gain API
        //

        _ = device.ListGains(direction, 0);

        _ = device.HasGainMode(direction, 0);
        _ = device.GetGainMode(direction, 0);
        device.SetGainMode(direction, 0, true);

        _ = device.GetGain(direction, 0);
        _ = device.GetGain(direction, 0, "");
        device.SetGain(direction, 0, 0.0);
        device.SetGain(direction, 0, "", 0.0);

        _ = device.GetGainRange(direction, 0);
        _ = device.GetGainRange(direction, 0, "");

        //
        // Frequency API
        //

        var frequencyArgsString = "key0=val0,key1=val1";

        var frequencyArgsMap = new Dictionary<string, string>();
        frequencyArgsMap["key0"] = "val0";
        frequencyArgsMap["key1"] = "val1";

        device.SetFrequency(direction, 0, 0.0);
        device.SetFrequency(direction, 0, 0.0, frequencyArgsString);
        device.SetFrequency(direction, 0, 0.0, frequencyArgsMap);
        _ = device.GetFrequency(direction, 0);

        device.SetFrequency(direction, 0, "", 0.0);
        device.SetFrequency(direction, 0, "", 0.0, frequencyArgsString);
        device.SetFrequency(direction, 0, "", 0.0, frequencyArgsMap);
        _ = device.GetFrequency(direction, 0, "");

        _ = device.ListFrequencies(direction, 0);

        _ = device.GetFrequencyRange(direction, 0);
        _ = device.GetFrequencyRange(direction, 0, "");
        _ = device.GetFrequencyArgsInfo(direction, 0);

        //
        // Sample rate API
        //

        device.SetSampleRate(direction, 0, 0.0);
        _ = device.GetSampleRate(direction, 0);
        _ = device.GetSampleRateRange(direction, 0);

        //
        // Bandwidth API
        //

        device.SetBandwidth(direction, 0, 0.0);
        _ = device.GetBandwidth(direction, 0);
        _ = device.GetBandwidthRange(direction, 0);

        //
        // Sensor API
        //

        _ = device.ListSensors(direction, 0);

        _ = device.ReadSensor(direction, 0, "");
        _ = device.ReadSensor<bool>(direction, 0, "");
        /*
        _ = device.ReadSensor<short>(direction, 0, "");
        _ = device.ReadSensor<int>(direction, 0, "");
        _ = device.ReadSensor<long>(direction, 0, "");
        _ = device.ReadSensor<ushort>(direction, 0, "");
        _ = device.ReadSensor<uint>(direction, 0, "");
        _ = device.ReadSensor<ulong>(direction, 0, "");
        _ = device.ReadSensor<float>(direction, 0, "");
        _ = device.ReadSensor<double>(direction, 0, "");
        */
        _ = device.ReadSensor<string>(direction, 0, "");

        //
        // Settings API
        //

        _ = device.GetSettingInfo(direction, 0);
        _ = device.GetSettingInfo(direction, 0, "");

        device.WriteSetting("", 0);
        device.WriteSetting("", 0.0);
        device.WriteSetting("", false);
        device.WriteSetting("", "");

        /*
        _ = device.ReadSetting(direction, 0, "");
        _ = device.ReadSetting<bool>(direction, 0, "");
        
        _ = device.ReadSetting<short>(direction, 0, "");
        _ = device.ReadSetting<int>(direction, 0, "");
        _ = device.ReadSetting<long>(direction, 0, "");
        _ = device.ReadSetting<ushort>(direction, 0, "");
        _ = device.ReadSetting<uint>(direction, 0, "");
        _ = device.ReadSetting<ulong>(direction, 0, "");
        _ = device.ReadSetting<float>(direction, 0, "");
        _ = device.ReadSetting<double>(direction, 0, "");
        
        _ = device.ReadSetting<string>(direction, 0, "");
        */
    }

    public static int Main(string[] args) => TestRunner.RunNUnitTest("TestDeviceAPI");
}